File: 2019/Javascript/day8.js
/*
 * ================================================
 * [ ADVENT OF CODE ] (https://adventofcode.com)
 * 2019 - Mina Pêcheux: Javascript (NodeJS) version
 * ------------------------------------------------
 * Day 8: Space Image Format
 * ================================================
 */
'use strict'

const _ = require('lodash')
const fs = require('fs')
const expect = require('chai').expect

// [ Computation functions ]
// -------------------------
/* Part I */
/**
 * Decodes the image layers by splitting it in even chunks.
 * @param {string} inputs - Content of the image (as a string).
 * @param {int} width - Width of each layer in the image.
 * @param {int} height - Height of each layer in the image.
 * @returns {array(string)} - Layers in the image.
 */
const decodeLayers = (inputs, width, height) => {
  const layerSize = width * height
  const nLayers = Math.round(inputs.length / layerSize)
  return _.map(_.range(nLayers), (l) => {
    return _.join(_.slice(inputs, l*layerSize, (l+1)*layerSize), '')
  })
}

/**
 * Computes a basic checksum to verify the image is intact by finding the layer
   that has the fewest layer and computing the product of its number of 1s and
   2s.
 * @param {string} inputs - Content of the image (as a string).
 * @param {int} width - Width of each layer in the image.
 * @param {int} height - Height of each layer in the image.
 * @returns {int} - Checksum to verify the image validity.
 */
const computeChecksum = (inputs, width, height) => {
  const layers = decodeLayers(inputs, width, height)
  let bestNZeros = null, bestValue = null, nZeros, nOnes, nTwos
  _.each(layers, (layer) => {
    nZeros = (layer.match(/0/g) || []).length
    if (bestNZeros === null || nZeros < bestNZeros) {
      nOnes = (layer.match(/1/g) || []).length
      nTwos = (layer.match(/2/g) || []).length
      bestNZeros = nZeros
      bestValue = nOnes * nTwos
    }
  })
  return { checksum: bestValue, layers }
}

/* Part II */
/**
 * Displays the message that was sent (and has previously been divided into
   even layers).
 * @param {array(string)} layers - Layers of the image.
 * @param {int} width - Width of each layer in the image.
 * @param {int} height - Height of each layer in the image.
 */
const displayMessage = (layers, width, height) => {
  const marker = '█'
  // iterate through layers in reverse order to have the right depht overwrite
  const img = _.map(_.range(height), () => new Array(width).fill(' '))
  _.each(_.reverse(layers), (layer) => {
    // go through grid
    _.each(_.range(height), (y) => {
      _.each(_.range(width), (x) => {
        // and turn on/off pixels depending on the value (transparent has no
        // impact)
        if (layer[x + y * width] === '1') {
          img[y][x] = marker
        } else if (layer[x + y * width] === '0') {
          img[y][x] = ' '
        }
      })
    })
  })
  // display the message
  console.log('')
  _.each(_.range(height), (y) => {
    console.log(_.join(img[y], ''))
  })
  console.log('')
}

// [ Base tests ]
// --------------
/**
 * Performs tests on the provided examples to check the result of the
 * computation functions is ok.
 */
const makeTests = () => {
  expect(decodeLayers('123456789012', 3, 2)).to.deep.equal([ '123456', '789012' ])
  expect(decodeLayers('210012011212', 3, 2)).to.deep.equal([ '210012', '011212' ])

  /* Part I */
  let tmp = computeChecksum('123456789012', 3, 2)
  expect(tmp.checksum).to.equal(1)
  tmp = computeChecksum('210012011212', 3, 2)
  expect(tmp.checksum).to.equal(6)
}

(() => {
  // check function results on example cases
  makeTests()
  
  // get input data
  const data = fs.readFileSync('../data/day8.txt')
  const inputs = data.toString()
  
  // Part I
  const solution1 = computeChecksum(inputs, 25, 6)
  console.log(`PART I: solution = ${solution1.checksum}`)
  
  // Part II
  const solution2 = displayMessage(solution1.layers, 25, 6)
  console.log('PART II: solution = (see the shell)')
})()

File: 2015/Python/day8.py
### =============================================
### [ ADVENT OF CODE ] (https://adventofcode.com)
### 2015 - Mina Pêcheux: Python version
### ---------------------------------------------
### Day 8: Matchsticks
### =============================================
import codecs
import re

# [ Input parsing functions ]
# ---------------------------
def parse_input(file):
    '''Parses the incoming file data into processable inputs.
    
    :param file: Open file pointer with the problem data.
    :type file: io.Stream
    :return: List of strings to check.
    :rtype: list(str)
    '''
    return [ line.strip() for line in file ]

# [ Computation functions ]
# -------------------------

### PART I
def decode_string(encoded):
    '''Decodes a string from its internal representation to the evaluated form.
    
    :param encoded: Encoded string (internal representation).
    :type encoded: str
    :return: Decoded version.
    :rtype: str
    '''
    return eval(encoded)

def get_decoding_diff(strings):
    '''Computes the total difference of length between the initial and the
    decoded (i.e. evaluated) strings.
    
    :param strings: List of strings to check.
    :type strings: list(str)
    :return: Difference of length between initial and decoded strings.
    :rtype: int
    '''
    s1 = 0
    s2 = 0
    for string in strings:
        decoded = decode_string(string)
        s1 += len(string)
        s2 += len(decoded)
    return s1 - s2

### PART II
def encode_string(decoded):
    '''Encodes a string from its classic representation to the escaped version.
    
    :param decoded: Decoded string.
    :type decoded: str
    :return: Escaped version of the string.
    :rtype: str
    '''
    return '"{}"'.format(re.escape(decoded))
    
def get_encoding_diff(strings):
    '''Computes the total difference of length between the encoded (i.e.
    escaped) and the initial strings.
    
    :param strings: List of strings to check.
    :type strings: list(str)
    :return: Difference of length between encoded and initial strings.
    :rtype: int
    '''
    s1 = 0
    s2 = 0
    for string in strings:
        encoded = encode_string(string)
        s1 += len(encoded)
        s2 += len(string)
    return s1 - s2
    
# [ Base tests ]
# --------------
def make_tests():
    '''Performs tests on the provided examples to check the result of the
    computation functions is ok.'''
    ### Part I
    assert decode_string('""') == ''
    assert decode_string('"abc"') == 'abc'
    assert decode_string('"aaa\\"aaa"') == 'aaa"aaa'
    assert decode_string('"\\x27"') == '\''
    
    ### Part II
    assert encode_string('""') == '"\\"\\""'
    assert encode_string('"abc"') == '"\\"abc\\""'
    assert encode_string('"aaa\\"aaa"') == '"\\"aaa\\\\\\"aaa\\""'
    assert encode_string('"\\x27"') == '"\\"\\\\x27\\""'
    
if __name__ == '__main__':
    # check function results on example cases
    make_tests()

    # get input data
    data_path = '../data/day8.txt'
    strings = parse_input(codecs.open(data_path, encoding='utf-8'))
    
    ### PART I
    solution = get_decoding_diff(strings)
    print('PART I: solution = {}'.format(solution))
    
    ### PART II
    solution = get_encoding_diff(strings)
    print('PART II: solution = {}'.format(solution))
